nixos/modules/services/security/kanidm.nix

```diff
@@ -199,10 +200,32 @@ in
         StateDirectory = "kanidm";
         StateDirectoryMode = "0700";
         ExecStart = "${pkgs.kanidm}/bin/kanidmd server -c ${serverConfigFile}";
-        ExecPreStart = ''
+        # There is currently no nice way to set the domain name, because origin
+        # in the config needs to be consistent with the current domain name:
+        # https://github.com/kanidm/kanidm/issues/623
+        ExecStartPre = pkgs.writeShellScript "kanidm-start-pre.sh" ''
+          set -euo pipefail
+          umask 077
+          log="$(${pkgs.coreutils}/bin/mktemp)"
+          # Try to set domain name
           ${pkgs.kanidm}/bin/kanidmd domain_name_change \
             -c ${serverConfigFile} \
-            -n ${cfg.ensureDomainName}
+            -n ${cfg.ensureDomainName} | ${pkgs.coreutils}/bin/tee "$log"
+          # Exit status is always 0, so search for error messages
+          if ${pkgs.gnugrep}/bin/grep -q 'admin.error' "$log"; then
+            # If it failed, extract the old domain name and create a temporary config
+            olddomain="$(${pkgs.gawk}/bin/awk 'match($0, /rp_id: "(.*)"/, m) { print m[1] }' "$log")"
+            ${pkgs.coreutils}/bin/rm "$log"
+
+            echo "Using temporary config with old domain name '$olddomain'"
+            # Create temporary config with old domain name
+            config="$(${pkgs.coreutils}/bin/mktemp)"
+            ${pkgs.gnused}/bin/sed ${dummyServerConfigFile} -e "s#@origin@#https://$olddomain#" > "$config"
+            ${pkgs.kanidm}/bin/kanidmd domain_name_change \
+              -c "$config" \
+              -n ${cfg.ensureDomainName}
+            ${pkgs.coreutils}/bin/rm "$config"
+          fi
         '';
```

Comment:
Using `systemd.services.kanidm.path = with pkgs; [ coreutils  gnugrep gnused ];` would make this script look less mutilated! :laughing: